File: python/biblioteca/libro.py
"""Class libro con las responsabilidades necesarias en la biblioteca"""

from genero import Genero


class Libro:

    generos = [
        Genero.INFANTILES,
        Genero.POESIA,
        Genero.CIENCIA_FICCION,
        Genero.AVENTURA,
        Genero.NOVELA,
        Genero.HISTORIA,
    ]

    def __init__(self, titulo, autor, genero, paginas):
        self.__titulo = titulo
        self.__autor = autor
        self.__set_genero(genero)
        self.__paginas = paginas

    def __set_genero(self, genero):
        if not genero in self.generos:
            raise ValueError("Genero incorrecto")
        self.__genero = genero

    def get_titulo(self) -> str:
        return self.__titulo

    def get_autor(self) -> str:
        return self.__autor

    def get_genero(self) -> Genero:
        return self.__genero

    def get_paginas(self) -> int:
        return self.__paginas

    def __str__(self) -> str:
        return (
            "Titulo: "
            + self.__titulo
            + " Autor: "
            + self.__autor
            + " Genero: "
            + str(self.__genero)
            + " Paginas: "
            + str(self.__paginas)
        )

    def __eq__(self, other) -> bool:
        return (
            self.__titulo == other.get_titulo()
            and self.__autor == other.get_autor()
            and self.get_genero() == other.get_genero()
            and self.get_paginas() == other.get_paginas()
        )

File: java/BibliotecaTN/Biblioteca/src/bibliotecaTN/Libro.java
package bibliotecaTN;

public class Libro {
    private String titulo;
    private String autor;
    private Genero genero;
    private int cantidadDePaginas;

    public Libro(String titulo, String autor, Genero genero, int cantidadDePaginas) {
        this.titulo = titulo;
        this.autor = autor;
        this.genero = genero;
        this.cantidadDePaginas = cantidadDePaginas;
    }

    public String getTitulo() {
        return titulo;
    }

    public String getAutor() {
        return autor;
    }

    public Genero getGenero() {
        return genero;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((titulo == null) ? 0 : titulo.hashCode());
        result = prime * result + ((autor == null) ? 0 : autor.hashCode());
        result = prime * result + ((genero == null) ? 0 : genero.hashCode());
        result = prime * result + cantidadDePaginas;
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        Libro other = (Libro) obj;
        if (titulo == null) {
            if (other.titulo != null)
                return false;
        } else if (!titulo.equals(other.titulo))
            return false;
        if (autor == null) {
            if (other.autor != null)
                return false;
        } else if (!autor.equals(other.autor))
            return false;
        if (genero != other.genero)
            return false;
        if (cantidadDePaginas != other.cantidadDePaginas)
            return false;
        return true;
    }

    @Override
    public String toString() {
        return "[" + titulo + ", " + autor + ", " + genero + ", cantidadDePaginas="
                + cantidadDePaginas + "]\n";
    }

    public int getCantidadDePaginas() {
        return cantidadDePaginas;
    }

}
